### PR TITLE
Add `fullwidth` messenger message

### PR DESCRIPTION
**Pull request by Jakeii**
## What does this change?

Add a messenger message to add a class that will set the slot to `width: 100%`

The class is added to DCR in this PR https://github.com/guardian/dotcom-rendering/pull/8917

## Why?
Usually front banner slots are restricted to the content area width (1300px), we would like our in house ad team and some ad partners to be able to have full-width ads in these slots.

Example usage from an ad:
```js
top.window.postMessage(JSON.stringify(
    {
        id: 'bf724866-723c-6b0a-e5d7-ad61535f98b7',
        type: 'full-width',
        value: true,
        source: 'clipcentric'

    }
), '*');
//send postMessage to commercial bundle
  top.window.postMessage(JSON.stringify(
    {
        id: 'bf724866-723c-6b0a-e5d7-ad61535f98b7',
        type: 'resize',
        value: {
        	width: '100vw',
         	height: '600px',
        },
      	source: 'clipcentric'
    }
), '*');
 
```

If the ad wanted to restrict to the column width again another message can be sent
```js
top.window.postMessage(JSON.stringify(
    {
        id: 'bf724866-723c-6b0a-e5d7-ad61535f98b7',
        type: 'full-width',
        value: false,
        source: 'clipcentric'

    }
), '*');

### CODE REVIEW

File: src/lib/messenger/fullwidth.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
+import { isObject } from '@guardian/libs';
+import fastdom from 'fastdom';
+import type { RegisterListener } from 'core/messenger';
+
+interface FullwidthSpecs {
+	fullwidth: boolean;
+}
+
+const isFullwidthSpecs = (specs: unknown): specs is FullwidthSpecs =>
+	!!specs && isObject(specs) && 'fullwidth' in specs;
```
Comment (chrislomaxjones) [inline src/lib/messenger/fullwidth.ts:10]: guardian/libs has got a `isBoolean` function if you wanted to validate the type of the `fullwidth` property as well

**Review by chrislomaxjones — APPROVED** [review]
Looks good

File: src/lib/messenger/fullwidth.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+			if (!isFullwidthSpecs(specs)) {
+				return;
+			}
+			const adSlot =
+				iframe.closest<HTMLElement>('.js-ad-slot') ?? undefined;
+
+			const name = adSlot?.dataset.name;
+
+			// only allow for banner ads
+			if (!name?.startsWith('fronts-banner') || !adSlot) {
```
Comment (domlander) [inline src/lib/messenger/fullwidth.ts:33]: if `adslot` is falsy then `name` will be undefined, so I don't think you need this second check

**Review by domlander — APPROVED** [review]
👍

File: src/core/messenger.ts [hunk 3]
```diff
@@ -13,6 +13,7 @@ type MessageType =
 	| 'measure-ad-load'
 	| 'passback'
 	| 'resize'
+	| 'fullwidth'
```
[the remark this answers is not in the document]
Comment (Jakeii) [reply src/core/messenger.ts:16]: ah thanks, I started changing it part way through, missed the message itself

File: src/lib/messenger/fullwidth.ts (hunk 2 above)
Comment (Jakeii) [reply src/lib/messenger/fullwidth.ts:33]: I agree it's a bit pointless but typescript complains without it when calling `fullwidth`

**Pull request merged**